contracts/payment/SponsorPayment.sol

```diff
@@ -0,0 +1,519 @@
+
+    // Event emitted when a withdrawal request is cancelled ERC20
+    event ERC20WithdrawalCancelled(address indexed sponsor, address indexed token, uint256 amount);
+
+    // Event emitted when a payment is claimed
+    event PaymentClaimed(address indexed recipient, uint256 indexed nonce, uint256 amount);
+
+    // Event emitted when a payment is claimed
+    event ERC20PaymentClaimed(
+        address indexed recipient,
+        uint256 indexed nonce,
+        address indexed token,
+        uint256 amount
+    );
+
+    // Event emitted when the owner withdraws their balance
+    event OwnerBalanceWithdrawn(uint256 amount);
+
+    // Event emitted when the owner withdraws their balance ERC20
+    event ERC20OwnerBalanceWithdrawn(uint256 amount);
+
+    /**
+     * @dev Payment details used in claim logic.
+     */
+    struct ERC20SponsorPaymentInfo {
+        address recipient;
+        uint256 amount;
+        address token;
+        uint256 nonce;
+        uint256 expiration;
+        bytes metadata;
+    }
+
+    /**
+     * @dev Payment details used in claim logic.
+     */
+    struct SponsorPaymentInfo {
+        address recipient;
+        uint256 amount;
+        uint256 nonce;
+        uint256 expiration;
+        bytes metadata;
+    }
+
+    struct WithdrawalRequest {
+        uint256 amount;
+        uint256 lockTime;
+        bool exists;
+    }
+
+    /**
+     * @dev Main storage $tructure for the contract
+     */
+    struct SponsorPaymentStorage {
+        mapping(address => mapping(address => uint256)) balances; // sponsor => token => balance
+        mapping(address => mapping(address => WithdrawalRequest)) withdrawalRequests; // sponsor => token => request
+        mapping(bytes32 => bool) isWithdrawn;
+        uint8 ownerPercentage;
+        uint256 withdrawalDelay;
+    }
+
+    string public constant VERSION = "1.0.0";
+
+    bytes32 public constant ERC20_PAYMENT_CLAIM_DATA_TYPE_HASH =
+        keccak256(
+            // solhint-disable-next-line max-line-length
+            "ERC20SponsorPaymentInfo(address recipient,uint256 amount,address token,uint256 expiration,uint256 nonce,bytes metadata)"
+        );
+
+    bytes32 public constant PAYMENT_CLAIM_DATA_TYPE_HASH =
+        keccak256(
+            // solhint-disable-next-line max-line-length
+            "ERC20SponsorPaymentInfo(address recipient,uint256 amount,uint256 expiration,uint256 nonce,bytes metadata)"
+        );
+
+    //  keccak256(abi.encode(uint256(keccak256("iden3.storage.SponsorPayment")) - 1)) &
+    //             ~bytes32(uint256(0xff));
+    bytes32 private constant SPONSOR_PAYMENT_STORAGE_LOCATION =
+        0x98fc76e32452055302f77aa95cd08aa0cf22c02a3ebdaee3e1411f6c47c2ef00;
+
+    modifier validToken(address token) {
+        if (token != address(0)) {
+            if (!_isContract(token)) revert InvalidToken("Not a contract address");
+            SponsorPaymentStorage storage s = _getSponsorPaymentStorage();
```

Comment:
do we need getting of storage in the modifier? why?

Reply:
fixed